<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core" />

<meta name="date" content="2022-11-16" />

<title>Module 11: DE Visualization and Annotation</title>

<script src="site_libs/header-attrs-2.17/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<script src="site_libs/navigation-1.1/sourceembed.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link rel="shortcut icon" href="favicon-16x16.png" />
<!--
Favicon from
https://twemoji.twitter.com/
https://favicon.io/emoji-favicons/dna/
-->

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>




<style type="text/css">
#rmd-source-code {
  display: none;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">RNA-Seq Demystified</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="Module00_Introduction.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Sequence QC
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module01_Warming_Up.html">Warming up</a>
    </li>
    <li>
      <a href="Module02a_Sequence_QC.html">Sequence QC</a>
    </li>
    <li>
      <a href="Module02b_Trimming.html">Trimming</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Alignment
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module03a_Reference_Genomes.html">Reference Genomes</a>
    </li>
    <li>
      <a href="Module03b_Alignment.html">Alignment</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Quantification
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module04_Alignment_QC_and_Quantification.html">Alignment QC and Quantification</a>
    </li>
    <li>
      <a href="Module05_Additional_Details.html">Optional: Additional details</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Groups/Model
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module06_DEAnalysisSetup.html">DE Introduction</a>
    </li>
    <li>
      <a href="Module07_DESeq2Init.html">DE Initialization</a>
    </li>
    <li>
      <a href="Module08_DESeq2DE.html">DE Normalization and Modeling</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Expression plots
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module09_SampleQCViz.html">Sample QC</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Diff Expression
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module10_DEComparisons.html">DE Testing</a>
    </li>
    <li>
      <a href="Module11_DEVisualizations.html">DE Visualization and Annotation</a>
    </li>
    <li>
      <a href="Module11X_BonusContent.html">Bonus content</a>
    </li>
  </ul>
</li>
<li>
  <a href="Module99_Wrap_up.html">Wrap up</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-download-source" href="#">Download Rmd</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Module 11: DE Visualization and
Annotation</h1>
<h4 class="author">UM Bioinformatics Core</h4>
<h4 class="date">2022-11-16</h4>

</div>


<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>
<blockquote>
<h1 id="objectives" class="unlisted unnumbered">Objectives</h1>
<ul>
<li>Understand advantages of using gene ids when analyzing data.</li>
<li>Given a list of ENSEMBL gene ids, add gene symbols and Entrez
accessions.</li>
<li>Generate common visualizations for differential expression
comparisons</li>
<li>Understand choosing thresholds for calling differentially expressed
genes</li>
<li>Discuss options for functional enrichments</li>
</ul>
</blockquote>
<div id="differential-expression-workflow"
class="section level1 unlisted unnumbered">
<h1 class="unlisted unnumbered">Differential Expression Workflow</h1>
<p>Here we will generate summary figures for our results and annotate
our DE tables.</p>
<p><img src="images/wayfinder/slides/Slide12.png"
style="width:75.0%" /></p>
<hr />
</div>
<div id="summarizing-de-results" class="section level1">
<h1>Summarizing DE results</h1>
<p>Part of differential expression analysis is generating visualizations
and summaries to share our results. While the DESeq2 tutorial provides
<a
href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#exploring-and-exporting-results">examples
of other visualizations</a>, a common visualization to summarize DE
comparisons are <a
href="http://resources.qiagenbioinformatics.com/manuals/clcgenomicsworkbench/752/index.php?manual=Volcano_plots_inspecting_result_statistical_analysis.html">volcano
plots</a>.</p>
<div id="tabular-de-summary" class="section level2">
<h2>Tabular DE summary</h2>
<p>To summarize DE genes, we first need thresholds for determining
significance. A reasonable threshold for linear fold-change is less than
-1.5 or greater than 1.5 (which corresponds to log2 fold-change -0.585
and 0.585, respectively. A standard threshold for the adjusted p-value
is less than 0.05.</p>
<p>Let’s set these as variables to reuse. This is generally good
practice because if these thresholds change upon later consideration,
you only have to change them in one location of your script, which will
reduce errors from missing some instances in your code.</p>
<pre class="r"><code>fc = 1.5
fdr = 0.05</code></pre>
<blockquote>
<h1 id="note-choosing-thresholds" class="unlisted unnumbered">Note:
Choosing thresholds</h1>
<p>Thresholding on adjusted p-values &lt; 0.05 is a standard threshold,
but depending on the research question and/or how the results will be
used, other thresholds might be reasonable.</p>
<p>There is a nice <a href="https://www.biostars.org/p/209118/">Biostar
post that discusses choosing adjusted p-value thresholds</a>, including
cases where a more relaxed threshold might be appropriate and (some
heated) discussion of the dangers of adjusting the choosen threshold
after running an analysis. Additionally, there is a <a
href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-13-S2-S11">Dalmon
et al 2012</a> paper about p-value and fold-change thresholds for
microarray data that may help provide some context.</p>
</blockquote>
<p>To generate a general summary of the DE results, we can use the
<code>summary</code> function to generate a basic summary by DESeq2.</p>
<pre class="r"><code>summary(results_minus_vs_plus)</code></pre>
<pre><code>
out of 16249 with nonzero total read count
adjusted p-value &lt; 0.1
LFC &gt; 0 (up)       : 53, 0.33%
LFC &lt; 0 (down)     : 38, 0.23%
outliers [1]       : 72, 0.44%
low counts [2]     : 3781, 23%
(mean count &lt; 7)
[1] see &#39;cooksCutoff&#39; argument of ?results
[2] see &#39;independentFiltering&#39; argument of ?results</code></pre>
<p>However, this summary is simply a text output that we are unable to
manipulate. Moreover, notice that the thresholds are not quite as we
would like them.</p>
<p>We can use conditional statements to determine the number of genes
that pass our thresholds for each comparison, which we can then use to
add information to the results table and plots.</p>
<blockquote>
<h1 id="exercise" class="unlisted unnumbered">Exercise</h1>
<p>How would we identify the number of genes with adjusted p-values &lt;
0.05 and a fold-change above 1.5 (or below -1.5)?</p>
</blockquote>
<details>
<summary>
Solution
</summary>
<p>Here is one possible answer:</p>
<pre class="r"><code>sum(results_minus_vs_plus$padj &lt; fdr &amp; abs(results_minus_vs_plus$log2FoldChange) &gt;= log2(fc), na.rm = TRUE)</code></pre>
<pre><code>[1] 56</code></pre>
</details>
<p><br></p>
<p>Let’s now create a new column in <code>results_minus_vs_plus</code>
to record the significance “call” based on these thresholds. And let’s
separate the call by “Up” or “Down”, noting that these are relative to
our “Case” condition of “plus”, or iron replete mice. There are many
ways to accomplish this, but the following will work:</p>
<p>First define all values as “NS” or “not significant”:</p>
<pre class="r"><code>results_minus_vs_plus$call = &#39;NS&#39;
head(results_minus_vs_plus)</code></pre>
<pre><code>log2 fold change (MLE): condition minus vs plus 
Wald test p-value: condition minus vs plus 
DataFrame with 6 rows and 7 columns
                      baseMean log2FoldChange     lfcSE      stat    pvalue
                     &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
ENSMUSG00000000001  1489.83039       0.297760  0.210310  1.415815  0.156830
ENSMUSG00000000028  1748.93544       0.226421  0.176795  1.280695  0.200301
ENSMUSG00000000031  2151.87715       0.457635  0.764579  0.598545  0.549476
ENSMUSG00000000037    24.91672       0.579130  0.561259  1.031840  0.302147
ENSMUSG00000000049     7.78377      -0.899483  1.553063 -0.579167  0.562476
ENSMUSG00000000056 19653.54030      -0.174048  0.203529 -0.855151  0.392468
                        padj        call
                   &lt;numeric&gt; &lt;character&gt;
ENSMUSG00000000001  0.868573          NS
ENSMUSG00000000028  0.902900          NS
ENSMUSG00000000031  0.995391          NS
ENSMUSG00000000037  0.950613          NS
ENSMUSG00000000049  0.998043          NS
ENSMUSG00000000056  0.982479          NS</code></pre>
<p>Next, determine the “Up” and “Down” indices:</p>
<pre class="r"><code>up_idx = results_minus_vs_plus$padj &lt; fdr &amp; results_minus_vs_plus$log2FoldChange &gt; log2(fc)
down_idx = results_minus_vs_plus$padj &lt; fdr &amp; results_minus_vs_plus$log2FoldChange &lt; -log2(fc)</code></pre>
<p>Last, use those indices to assign the correct “Up” or “Down” values
to the correct indices, and look at the head of the result:</p>
<pre class="r"><code>results_minus_vs_plus$call[up_idx] = &#39;Up&#39;
results_minus_vs_plus$call[down_idx] = &#39;Down&#39;
head(results_minus_vs_plus)</code></pre>
<pre><code>log2 fold change (MLE): condition minus vs plus 
Wald test p-value: condition minus vs plus 
DataFrame with 6 rows and 7 columns
                      baseMean log2FoldChange     lfcSE      stat    pvalue
                     &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
ENSMUSG00000000001  1489.83039       0.297760  0.210310  1.415815  0.156830
ENSMUSG00000000028  1748.93544       0.226421  0.176795  1.280695  0.200301
ENSMUSG00000000031  2151.87715       0.457635  0.764579  0.598545  0.549476
ENSMUSG00000000037    24.91672       0.579130  0.561259  1.031840  0.302147
ENSMUSG00000000049     7.78377      -0.899483  1.553063 -0.579167  0.562476
ENSMUSG00000000056 19653.54030      -0.174048  0.203529 -0.855151  0.392468
                        padj        call
                   &lt;numeric&gt; &lt;character&gt;
ENSMUSG00000000001  0.868573          NS
ENSMUSG00000000028  0.902900          NS
ENSMUSG00000000031  0.995391          NS
ENSMUSG00000000037  0.950613          NS
ENSMUSG00000000049  0.998043          NS
ENSMUSG00000000056  0.982479          NS</code></pre>
<p>Finally, looking ahead to when we plot these values as colors in a
volcano plot, let’s make this <code>call</code> column a factor and
specify the level ordering:</p>
<pre class="r"><code>results_minus_vs_plus$call = factor(results_minus_vs_plus$call, levels = c(&#39;Up&#39;, &#39;Down&#39;, &#39;NS&#39;))
head(results_minus_vs_plus)</code></pre>
<pre><code>log2 fold change (MLE): condition minus vs plus 
Wald test p-value: condition minus vs plus 
DataFrame with 6 rows and 7 columns
                      baseMean log2FoldChange     lfcSE      stat    pvalue
                     &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
ENSMUSG00000000001  1489.83039       0.297760  0.210310  1.415815  0.156830
ENSMUSG00000000028  1748.93544       0.226421  0.176795  1.280695  0.200301
ENSMUSG00000000031  2151.87715       0.457635  0.764579  0.598545  0.549476
ENSMUSG00000000037    24.91672       0.579130  0.561259  1.031840  0.302147
ENSMUSG00000000049     7.78377      -0.899483  1.553063 -0.579167  0.562476
ENSMUSG00000000056 19653.54030      -0.174048  0.203529 -0.855151  0.392468
                        padj     call
                   &lt;numeric&gt; &lt;factor&gt;
ENSMUSG00000000001  0.868573       NS
ENSMUSG00000000028  0.902900       NS
ENSMUSG00000000031  0.995391       NS
ENSMUSG00000000037  0.950613       NS
ENSMUSG00000000049  0.998043       NS
ENSMUSG00000000056  0.982479       NS</code></pre>
<blockquote>
<h1 id="tip" class="unlisted unnumbered">Tip</h1>
<p>It is often helpful to include code like this in differential
expression analyses so there is a clearly labelled column that makes
subsetting and summarizing the results easier.</p>
</blockquote>
<p>Now we are in a position to quickly summarize our differential
expression results:</p>
<pre class="r"><code>table(results_minus_vs_plus$call)</code></pre>
<pre><code>
   Up  Down    NS 
   41    15 16193 </code></pre>
<p>We see quickly how many genes were “Up” in iron replete, how many
were “Down” in iron replete, and how many were not significant.</p>
<p><strong>Checkpoint</strong>: <em>If you successfully added the
<code>call</code> column and got the same table result as above, please
indicate with a green check. Otherwise use a red x.</em></p>
</div>
<div id="visual-de-summary" class="section level2">
<h2>Visual DE summary</h2>
<p>As described by this <a
href="https://galaxyproject.github.io/training-material/topics/transcriptomics/tutorials/rna-seq-viz-with-volcanoplot/tutorial.html">Galaxy
project tutorial</a>, a volcano plot is a type of scatterplot that shows
statistical significance (adjusted p-value) versus magnitude of change
(fold change). In a volcano plot, the most upregulated genes are towards
the right, the most downregulated genes are towards the left, and the
most statistically significant genes are towards the top.</p>
<p>Let’s coerce the <code>DataFrame</code> which was returned by
<code>DESeq2::results()</code> into a <code>tibble</code> in
anticipation of using the <code>ggplot2</code> library to plot. We’re
also going to modify our results table so that the row names become a
separate column, and so that it’s ordered by adjusted p-value.</p>
<pre class="r"><code># Use the rownames argument to create a new column of gene IDs
# Also arrange by adjusted p-value
results_minus_vs_plus = as_tibble(results_minus_vs_plus, rownames = &#39;id&#39;) %&gt;% arrange(padj)</code></pre>
<p>Let’s start with a very simple volcano plot that plots the
<code>log2FoldChange</code> on the x-axis, and <code>-log10(padj)</code>
on the y-axis.</p>
<pre class="r"><code># Initialize the plot, saving as object &#39;p&#39; and specifying the plot type as &#39;geom_point&#39;
p = ggplot(results_minus_vs_plus, aes(x = log2FoldChange, y = -log10(padj))) +
    geom_point()
p</code></pre>
<pre><code>Warning: Removed 3853 rows containing missing values (`geom_point()`).</code></pre>
<p><img src="fig/11-VolcanoPlot-1.png" width="816" style="display: block; margin: auto;" /></p>
<p>This is a good start, but, as usual it’s nice to add better labels to
the plot with the <code>labs()</code> function:</p>
<pre class="r"><code># Add plot labels and change the theme
p = ggplot(results_minus_vs_plus, aes(x = log2FoldChange, y = -log10(padj))) +
    geom_point() +
    theme_bw() +
    labs(
        title = &#39;Volcano Plot&#39;,
        subtitle = &#39;Plus vs Minus&#39;,
        x = &#39;log2 fold-change&#39;,
        y = &#39;-log10 FDR&#39;
    )
p</code></pre>
<pre><code>Warning: Removed 3853 rows containing missing values (`geom_point()`).</code></pre>
<p><img src="fig/11-VolcanoPlot2-1.png" width="816" style="display: block; margin: auto;" /></p>
<p>This is a nice improvement. What if we now added some visual guides
to indicate where the significant genes are? We can uset he
<code>geom_vline()</code> and <code>geom_hline()</code> functions to
accomplish this:</p>
<pre class="r"><code># Add threshold lines
p = p +
    geom_vline(
        xintercept = c(0, -log2(fc), log2(fc)),
        linetype = c(1, 2, 2)) +
    geom_hline(
        yintercept = -log10(fdr),
        linetype = 2)
p</code></pre>
<pre><code>Warning: Removed 3853 rows containing missing values (`geom_point()`).</code></pre>
<p><img src="fig/11-VolcanoPlot3-1.png" width="816" style="display: block; margin: auto;" /></p>
<p>Finally, why not color the points by their significance status? We
already created the <code>call</code> column that has the correct
values. In this case we can get away with adding a last
<code>geom_point()</code> and specifying the correct aesthetic:</p>
<pre class="r"><code>p = p + geom_point(aes(color = call))
p</code></pre>
<pre><code>Warning: Removed 3853 rows containing missing values (`geom_point()`).
Removed 3853 rows containing missing values (`geom_point()`).</code></pre>
<p><img src="fig/11-VolcanoPlot4-1.png" width="816" style="display: block; margin: auto;" /></p>
<p>For additional visualizations for our DE results, we included some
example code in the Bonus Content module and this <a
href="https://hbctraining.github.io/DGE_workshop/lessons/06_DGE_visualizing_results.html">HBC
tutorial</a> includes some nice examples.</p>
<div id="subsetting-significant-genes" class="section level3">
<h3>Subsetting significant genes</h3>
<p>You may be interested in identifying only the genes that pass your
significance thresholds. A useful way to do this is to conditionally
subset your results. Again, we already created the <code>call</code>
column, which makes this relatively simple to do.</p>
<p><em>Note: The tidyverse functions you learned in Software Carpentry
could also be alternatively used here.</em></p>
<pre class="r"><code>res_sig = results_minus_vs_plus[results_minus_vs_plus$call != &#39;NS&#39;, ]
head(res_sig)</code></pre>
<pre><code># A tibble: 6 × 8
  id                 baseMean log2FoldChange lfcSE  stat   pvalue     padj call 
  &lt;chr&gt;                 &lt;dbl&gt;          &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt; &lt;fct&gt;
1 ENSMUSG00000032715   1178.           3.64  0.450  8.10 5.71e-16 7.08e-12 Up   
2 ENSMUSG00000101645     77.3        -10.1   1.33  -7.58 3.42e-14 2.12e-10 Down 
3 ENSMUSG00000027313    243.           2.67  0.410  6.50 7.96e-11 2.47e- 7 Up   
4 ENSMUSG00000027889   3250.           0.946 0.145  6.52 7.00e-11 2.47e- 7 Up   
5 ENSMUSG00000054136     92.6          3.39  0.526  6.44 1.19e-10 2.95e- 7 Up   
6 ENSMUSG00000020142    625.           1.72  0.287  5.99 2.07e- 9 4.28e- 6 Up   </code></pre>
<pre class="r"><code>dim(res_sig)</code></pre>
<pre><code>[1] 56  8</code></pre>
</div>
</div>
<div id="adding-genome-annotations" class="section level2">
<h2>Adding genome annotations</h2>
<p>Since, gene symbols can change over time or be ambiguous we use, and
recommend, using the EMSEMBL reference genome and ENSEMBL IDs for
alignments and we’ve been working with tables and data where all genes
are labeled only by their long ENSEMBL ID (you will notice this in the
GTF we used in the reference). However, this can make it difficult to
quickly look at results for genes of interest.</p>
<p>Luckily, Bioconductor provides many tools and resources to facilitate
access to <a
href="http://bioconductor.org/packages/devel/workflows/vignettes/annotation/inst/doc/Annotation_Resources.html">genomic
annotation resources</a>.</p>
<p>To start, we will first load the <a
href="https://bioconductor.org/packages/3.14/bioc/html/biomaRt.html">biomaRt
library</a> and choose what reference we want to access. For a more
detailed walk through of using biomaRt, <a
href="https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2019/RNAseq/html/05_Annotation_and_Visualisation.html">this
training module</a> might be useful, including what to do when
annotations are not 1:1 mappings.</p>
<p>We’ll start by loading the <code>biomaRt</code> library and calling
the <code>useEnsembl()</code> function to select the database we’ll use
to extract the information we need. This will download the mapping of
ENSEMBL IDs to gene symbols, enabling us to eventually add the gene
symbol column we want.</p>
<pre class="r"><code>library(&#39;biomaRt&#39;)
ensembl = useEnsembl(dataset = &#39;mmusculus_gene_ensembl&#39;, biomart=&#39;ensembl&#39;)</code></pre>
<p><em>Note that this process takes some time and will take up a larger
amount of working memory so proceed with caution if you try to run these
commands on a laptop with less than 4G of memory</em></p>
<p>To identify possible <strong>filters</strong> to restrict our data,
we can use the <code>listFilters</code> function. To identify the
<strong>attributes</strong> we want to retrive, we can use the
<code>listAttributes</code> function. The best approach is to use <a
href="https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html#how-to-build-a-biomart-query">list
or search functions</a> to help narrow down the available options.</p>
<pre class="r"><code>head(listFilters(mart = ensembl), n = 20)
head(listAttributes(ensembl), n = 30)</code></pre>
<p>We can access additional genomic annotations using the <a
href="https://bioconductor.org/packages/release/bioc/html/biomaRt.html"><code>bioMart</code>
package</a>. To identify we’ll structure our ‘query’ or search of the
bioMart resources to use the <a
href="https://m.ensembl.org/info/genome/genebuild/gene_names.html">ENSEMBL
id</a> from our alignment to add the gene symbols and gene description
for each gene.</p>
<pre class="r"><code>id_mapping = getBM(attributes=c(&#39;ensembl_gene_id&#39;, &#39;external_gene_name&#39;),
      filters = &#39;ensembl_gene_id&#39;,
      values = row.names(assay(dds)),
      mart = ensembl)
# will take some time to run

# Preview the result
head(id_mapping)</code></pre>
<pre><code>     ensembl_gene_id external_gene_name
1 ENSMUSG00000000001              Gnai3
2 ENSMUSG00000000028              Cdc45
3 ENSMUSG00000000031                H19
4 ENSMUSG00000000037              Scml2
5 ENSMUSG00000000049               Apoh
6 ENSMUSG00000000056               Narf</code></pre>
<p>Now that we have the ENSEMBL information and a gene symbol to match
to our results, we can proceed in the smaller groups. As with the
previous exercise, we have broken it into small steps with hints as
needed.</p>
<p><strong>Note</strong>: For additional information regarding bioMart,
please consult the <a
href="https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html">ENSEMBL
bioMart vignette</a> or the broader <a
href="http://bioconductor.org/packages/devel/workflows/vignettes/annotation/inst/doc/Annotation_Resources.html">Bioconductor
Annotation Resources vignette</a>.</p>
<p>Look at the two data frames that are going to be needed:
<code>id_mapping</code> and <code>results_minus_vs_plus</code>.</p>
<pre class="r"><code>head(id_mapping)</code></pre>
<pre><code>     ensembl_gene_id external_gene_name
1 ENSMUSG00000000001              Gnai3
2 ENSMUSG00000000028              Cdc45
3 ENSMUSG00000000031                H19
4 ENSMUSG00000000037              Scml2
5 ENSMUSG00000000049               Apoh
6 ENSMUSG00000000056               Narf</code></pre>
<pre class="r"><code>head(results_minus_vs_plus)</code></pre>
<pre><code># A tibble: 6 × 8
  id                 baseMean log2FoldChange lfcSE  stat   pvalue     padj call 
  &lt;chr&gt;                 &lt;dbl&gt;          &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt; &lt;fct&gt;
1 ENSMUSG00000032715   1178.           3.64  0.450  8.10 5.71e-16 7.08e-12 Up   
2 ENSMUSG00000101645     77.3        -10.1   1.33  -7.58 3.42e-14 2.12e-10 Down 
3 ENSMUSG00000027313    243.           2.67  0.410  6.50 7.96e-11 2.47e- 7 Up   
4 ENSMUSG00000027889   3250.           0.946 0.145  6.52 7.00e-11 2.47e- 7 Up   
5 ENSMUSG00000054136     92.6          3.39  0.526  6.44 1.19e-10 2.95e- 7 Up   
6 ENSMUSG00000020142    625.           1.72  0.287  5.99 2.07e- 9 4.28e- 6 Up   </code></pre>
<p>We want to match the <code>id</code> column of
<code>results_minus_vs_plus</code> to the <code>ensembl_gene_id</code>
column of <code>id_mapping</code>, and once that match is found, we want
to extract the <code>external_gene_name</code> column of
<code>id_mapping</code> to get the gene symbol. Next, look at the
documentation for <code>dplyr::left_join()</code> and merge the
<code>id_mapping</code> table into the
<code>results_minus_vs_plus</code> table on the columns
<code>ensembl_gene_id</code> and <code>external_gene_name</code>.</p>
<pre class="r"><code>results_minus_vs_plus_annotated = results_minus_vs_plus %&gt;%
    left_join(id_mapping, by = c(&#39;id&#39; = &#39;ensembl_gene_id&#39;))
head(results_minus_vs_plus_annotated)</code></pre>
<pre><code># A tibble: 6 × 9
  id                 baseM…¹ log2F…² lfcSE  stat   pvalue     padj call  exter…³
  &lt;chr&gt;                &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt; &lt;fct&gt; &lt;chr&gt;  
1 ENSMUSG00000032715  1178.    3.64  0.450  8.10 5.71e-16 7.08e-12 Up    Trib3  
2 ENSMUSG00000101645    77.3 -10.1   1.33  -7.58 3.42e-14 2.12e-10 Down  Gm28635
3 ENSMUSG00000027313   243.    2.67  0.410  6.50 7.96e-11 2.47e- 7 Up    Chac1  
4 ENSMUSG00000027889  3250.    0.946 0.145  6.52 7.00e-11 2.47e- 7 Up    Ampd2  
5 ENSMUSG00000054136    92.6   3.39  0.526  6.44 1.19e-10 2.95e- 7 Up    Adm2   
6 ENSMUSG00000020142   625.    1.72  0.287  5.99 2.07e- 9 4.28e- 6 Up    Slc1a4 
# … with abbreviated variable names ¹​baseMean, ²​log2FoldChange,
#   ³​external_gene_name</code></pre>
<p>We can use some of the <code>tidyverse</code> functions we’ve
encountered previously to rename the <code>external_gene_name</code>
column to <code>symbol</code> and to move it into the second column
position? Hint: Because of the order of the packages we may have loaded,
we’ll use <code>dplyr::rename()</code> and <code>dplyr::select()</code>
instead of just the <code>select()</code> function. We can discuss this
in a moment.</p>
<pre class="r"><code>results_minus_vs_plus_annotated = results_minus_vs_plus_annotated %&gt;%
    dplyr::rename(&#39;symbol&#39; = &#39;external_gene_name&#39;) %&gt;%
    dplyr::select(id, symbol, everything())
results_minus_vs_plus_annotated</code></pre>
<pre><code># A tibble: 16,249 × 9
   id                 symbol baseM…¹ log2F…² lfcSE  stat   pvalue     padj call 
   &lt;chr&gt;              &lt;chr&gt;    &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt; &lt;fct&gt;
 1 ENSMUSG00000032715 Trib3   1178.    3.64  0.450  8.10 5.71e-16 7.08e-12 Up   
 2 ENSMUSG00000101645 Gm286…    77.3 -10.1   1.33  -7.58 3.42e-14 2.12e-10 Down 
 3 ENSMUSG00000027313 Chac1    243.    2.67  0.410  6.50 7.96e-11 2.47e- 7 Up   
 4 ENSMUSG00000027889 Ampd2   3250.    0.946 0.145  6.52 7.00e-11 2.47e- 7 Up   
 5 ENSMUSG00000054136 Adm2      92.6   3.39  0.526  6.44 1.19e-10 2.95e- 7 Up   
 6 ENSMUSG00000020142 Slc1a4   625.    1.72  0.287  5.99 2.07e- 9 4.28e- 6 Up   
 7 ENSMUSG00000040280 Ndufa…    85.9   3.62  0.628  5.75 8.68e- 9 1.54e- 5 Up   
 8 ENSMUSG00000023951 Vegfa    350.    2.44  0.432  5.64 1.67e- 8 2.59e- 5 Up   
 9 ENSMUSG00000115420 Rmrp      16.0 -21.9   3.91  -5.60 2.11e- 8 2.90e- 5 Down 
10 ENSMUSG00000027737 Slc7a…   119.    2.32  0.429  5.41 6.47e- 8 8.02e- 5 Up   
# … with 16,239 more rows, and abbreviated variable names ¹​baseMean,
#   ²​log2FoldChange</code></pre>
<p>And now we have our differential expression results annotated with
gene symbols, which can help in the interpretation of the results, and
can be used in downstream analysis such as functional analysis.</p>
</div>
</div>
<div id="outputting-results-to-file" class="section level1">
<h1>Outputting results to file</h1>
<p>A key aspect of our analysis is preserving the relevant datasets for
both our records and for downstream applications, such as functional
enrichments.</p>
<div id="de-results-table" class="section level2">
<h2>DE results table</h2>
<p>Next we’ll write out our DE results, now that we’ve added information
to the table to help us interpret the results, and share with
collaborators.</p>
<pre class="r"><code>write.csv(results_minus_vs_plus,
          row.names = FALSE,
          na = &quot;.&quot;,
          file=&quot;outputs/tables/DE_results_minus_vs_plus.csv&quot;)</code></pre>
</div>
<div id="count-tables" class="section level2">
<h2>Count tables</h2>
<p>The most relevant count tables are the raw, filtered count table that
we used as the input for our analysis and the rlog normalized count
table that we used for our quality control visualizations.</p>
<p>To output the raw counts, we will need to use the <code>counts</code>
function to access the count table from within its larger
<code>DESeqDataSet</code> object.</p>
<pre class="r"><code>write.csv(counts(dds, normalized = FALSE), file=&quot;outputs/tables/DESeq2_raw_counts.csv&quot;)</code></pre>
<p>Then we’ll output the rlog count table, using the <code>assay</code>
function to access the normalized count table from within its larger
<code>DESeqDataSet</code> object.</p>
<pre class="r"><code>write.csv(assay(rld), file=&quot;outputs/tables/DESeq2_rlog_normalized_counts.csv&quot;)</code></pre>
</div>
</div>
<div id="summary" class="section level1">
<h1>Summary</h1>
<p>In this section, we:</p>
<ul>
<li>Generated a volcano plot for our differential expression
results</li>
<li>Summarized our differential expression results</li>
<li>Discussed choosing thresholds</li>
<li>Annotated our tables of results to map gene IDs to gene symbols</li>
<li>Saved our results to file</li>
</ul>
</div>
<div id="key-takeaways" class="section level1">
<h1>Key takeaways</h1>
<p>Overall, we’ve run through most of the building blocks needed to run
a differential expression analysis and hopefully built up a better
understanding of how differential expression comparisons work,
particularly how experimental design can impact our results.</p>
<p>What to consider moving forward:</p>
<ul>
<li>How can I control for technical variation in my experimental
design?</li>
<li>How much variation is expected with a treatment group?</li>
<li>What is my RNA quality, and how can that be optimized?</li>
<li>Are there quality concerns for my sequencing data?</li>
<li>What comparisons are relevant to my biological question?</li>
<li>Are there covariates that should be considered?</li>
<li>What will a differential expression analysis tell me?</li>
</ul>
<p><strong>Let’s pause here for general questions</strong></p>
<hr />
</div>
<div id="what-can-we-do-with-our-de-results" class="section level1">
<h1>What can we do with our DE results?</h1>
<p>A way to determine possible <a
href="https://www.ebi.ac.uk/training-beta/online/courses/functional-genomics-ii-common-technologies-and-data-analysis-methods/biological-interpretation-of-gene-expression-data-2/">broader
biological interpretations</a> from the observed DE results, is
functional enrichments. There are many options, such as some included in
this <a
href="https://www.researchgate.net/post/How_can_I_analyze_a_set_of_DEGs_differentially_expressed_genes_to_obtain_information_from_them">discussion
thread</a>. Other common functional enrichments approaches are gene set
enrichment analysis, aka <a
href="http://software.broadinstitute.org/gsea/index.jsp">GSEA</a>,
Database for Annotation, Visualization and Integrated Discovery, aka <a
href="https://david.ncifcrf.gov/">DAVID</a>, <a
href="https://digitalinsights.qiagen.com/">Ingenity</a>, and [iPathway
Guide]</p>
<p>The University of Michigan has license and support for additional
tools, such as Cytoscape, so we recommend reaching out to staff with <a
href="https://www.lib.umich.edu/locations-and-hours/taubman-health-sciences-library/research-and-clinical-support">Taubman
Library</a> to learn more about resources that might be application
toyour research.</p>
<hr />
</div>
<div id="sources" class="section level1">
<h1>Sources</h1>
<ul>
<li>HBC DGE training module, part 1: <a
href="https://hbctraining.github.io/DGE_workshop/lessons/04_DGE_DESeq2_analysis.html"
class="uri">https://hbctraining.github.io/DGE_workshop/lessons/04_DGE_DESeq2_analysis.html</a></li>
<li>HBC DGE training module, part 2: <a
href="https://hbctraining.github.io/DGE_workshop/lessons/05_DGE_DESeq2_analysis2.html"
class="uri">https://hbctraining.github.io/DGE_workshop/lessons/05_DGE_DESeq2_analysis2.html</a></li>
<li>DESeq2 vignette: <a
href="http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#differential-expression-analysis"
class="uri">http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#differential-expression-analysis</a></li>
<li>Bioconductor Genomic Annotation resources: <a
href="http://bioconductor.org/packages/devel/workflows/vignettes/annotation/inst/doc/Annotation_Resources.html"
class="uri">http://bioconductor.org/packages/devel/workflows/vignettes/annotation/inst/doc/Annotation_Resources.html</a></li>
<li>BioMart vignette: <a
href="https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html"
class="uri">https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html</a></li>
</ul>
</div>
<div id="additional-resources" class="section level1">
<h1>Additional Resources</h1>
<ul>
<li>MIDAS Reproduciblity Hub: <a
href="https://midas.umich.edu/reproducibility-overview/"
class="uri">https://midas.umich.edu/reproducibility-overview/</a></li>
<li>ARC resources: <a href="https://arc-ts.umich.edu/"
class="uri">https://arc-ts.umich.edu/</a></li>
<li>Gene Set Enrichment Resources from Bioconductor: <a
href="https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2018/RNASeq2018/html/06_Gene_set_testing.nb.html"
class="uri">https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2018/RNASeq2018/html/06_Gene_set_testing.nb.html</a></li>
<li>Using HTSeq data with DESeq2: <a
href="https://angus.readthedocs.io/en/2019/diff-ex-and-viz.html"
class="uri">https://angus.readthedocs.io/en/2019/diff-ex-and-viz.html</a></li>
<li>Detailed RNA-seq analysis paper: <a
href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6096346/"
class="uri">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6096346/</a></li>
<li>Overview of RNA-seq analysis considerations: <a
href="https://academic-oup-com.proxy.lib.umich.edu/bfg/article/14/2/130/257370"
class="uri">https://academic-oup-com.proxy.lib.umich.edu/bfg/article/14/2/130/257370</a></li>
<li>Alternative overview of DESeq2, including visualizations and
functional enrichments: <a
href="http://dputhier.github.io/jgb71e-polytech-bioinfo-app/practical/rna-seq_R/rnaseq_diff_Snf2.html"
class="uri">http://dputhier.github.io/jgb71e-polytech-bioinfo-app/practical/rna-seq_R/rnaseq_diff_Snf2.html</a></li>
</ul>
<hr />
</div>
<div id="session-info" class="section level1">
<h1>Session Info</h1>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.2.1 (2022-06-23)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Big Sur ... 10.16

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] forcats_0.5.2               stringr_1.4.1              
 [3] purrr_0.3.5                 readr_2.1.3                
 [5] tibble_3.1.8                tidyverse_1.3.2            
 [7] biomaRt_2.52.0              data.table_1.14.4          
 [9] RColorBrewer_1.1-3          pheatmap_1.0.12            
[11] ggrepel_0.9.2               dplyr_1.0.10               
[13] tidyr_1.2.1                 ggplot2_3.4.0              
[15] DESeq2_1.36.0               SummarizedExperiment_1.26.1
[17] Biobase_2.56.0              MatrixGenerics_1.8.1       
[19] matrixStats_0.62.0          GenomicRanges_1.48.0       
[21] GenomeInfoDb_1.32.4         IRanges_2.30.1             
[23] S4Vectors_0.34.0            BiocGenerics_0.42.0        
[25] knitr_1.40                  rmarkdown_2.17             

loaded via a namespace (and not attached):
 [1] googledrive_2.0.0      colorspace_2.0-3       ellipsis_0.3.2        
 [4] XVector_0.36.0         fs_1.5.2               farver_2.1.1          
 [7] bit64_4.0.5            AnnotationDbi_1.58.0   fansi_1.0.3           
[10] lubridate_1.9.0        xml2_1.3.3             codetools_0.2-18      
[13] splines_4.2.1          cachem_1.0.6           geneplotter_1.74.0    
[16] jsonlite_1.8.3         broom_1.0.1            annotate_1.74.0       
[19] dbplyr_2.2.1           png_0.1-7              compiler_4.2.1        
[22] httr_1.4.4             backports_1.4.1        assertthat_0.2.1      
[25] Matrix_1.5-1           fastmap_1.1.0          gargle_1.2.1          
[28] cli_3.4.1              htmltools_0.5.3        prettyunits_1.1.1     
[31] tools_4.2.1            gtable_0.3.1           glue_1.6.2            
[34] GenomeInfoDbData_1.2.8 rappdirs_0.3.3         Rcpp_1.0.9            
[37] cellranger_1.1.0       jquerylib_0.1.4        vctrs_0.5.0           
[40] Biostrings_2.64.1      xfun_0.34              rvest_1.0.3           
[43] timechange_0.1.1       lifecycle_1.0.3        XML_3.99-0.12         
[46] googlesheets4_1.0.1    zlibbioc_1.42.0        scales_1.2.1          
[49] hms_1.1.2              parallel_4.2.1         yaml_2.3.6            
[52] curl_4.3.3             memoise_2.0.1          sass_0.4.2            
[55] stringi_1.7.8          RSQLite_2.2.18         highr_0.9             
[58] genefilter_1.78.0      filelock_1.0.2         BiocParallel_1.30.4   
[61] rlang_1.0.6            pkgconfig_2.0.3        bitops_1.0-7          
[64] evaluate_0.18          lattice_0.20-45        labeling_0.4.2        
[67] bit_4.0.4              tidyselect_1.2.0       magrittr_2.0.3        
[70] R6_2.5.1               generics_0.1.3         DelayedArray_0.22.0   
[73] DBI_1.1.3              pillar_1.8.1           haven_2.5.1           
[76] withr_2.5.0            survival_3.4-0         KEGGREST_1.36.3       
[79] RCurl_1.98-1.9         modelr_0.1.9           crayon_1.5.2          
[82] utf8_1.2.2             BiocFileCache_2.4.0    tzdb_0.3.0            
[85] progress_1.2.2         readxl_1.4.1           locfit_1.5-9.6        
[88] grid_4.2.1             blob_1.2.3             reprex_2.0.2          
[91] digest_0.6.30          xtable_1.8-4           munsell_0.5.0         
[94] bslib_0.4.1           </code></pre>
<hr />
<p>These materials have been adapted and extended from materials listed
above. These are open access materials distributed under the terms of
the <a href="http://creativecommons.org/licenses/by/4.0/">Creative
Commons Attribution license (CC BY 4.0)</a>, which permits unrestricted
use, distribution, and reproduction in any medium, provided the original
author and source are credited.</p>
</div>

<div id="rmd-source-code">---
title: "Module 11: DE Visualization and Annotation"
author: "UM Bioinformatics Core"
date: "`r Sys.Date()`"
output:
        html_document:
            includes:
                in_header: header.html
            theme: paper
            toc: true
            toc_depth: 4
            toc_float: true
            number_sections: false
            fig_caption: true
            markdown: GFM
            code_download: true
---

<style type="text/css">
body, td {
   font-size: 18px;
}
code.r{
  font-size: 12px;
}
pre {
  font-size: 12px
}
</style>

```{r, include = FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("11-")
```

> # Objectives {.unlisted .unnumbered}
> * Understand advantages of using gene ids when analyzing data.
> * Given a list of ENSEMBL gene ids, add gene symbols and Entrez accessions.
> * Generate common visualizations for differential expression comparisons
> * Understand choosing thresholds for calling differentially expressed genes
> * Discuss options for functional enrichments

```{r Modules, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
library(DESeq2)
library(ggplot2)
library(tidyr)
library(dplyr)
library(matrixStats)
library(ggrepel)
library(pheatmap)
library(RColorBrewer)
# load("rdata/RunningData.RData")
```

# Differential Expression Workflow {.unlisted .unnumbered}

Here we will generate summary figures for our results and annotate our DE tables.

![](./images/wayfinder/wayfinder-DEVisualizations.png){width=75%}

---

# Summarizing DE results

Part of differential expression analysis is generating visualizations and summaries to share our results. While the DESeq2 tutorial provides [examples of other visualizations](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#exploring-and-exporting-results), a common visualization to summarize DE comparisons are [volcano plots](http://resources.qiagenbioinformatics.com/manuals/clcgenomicsworkbench/752/index.php?manual=Volcano_plots_inspecting_result_statistical_analysis.html).

## Tabular DE summary

To summarize DE genes, we first need thresholds for determining significance. A reasonable threshold for linear fold-change is less than -1.5 or greater than 1.5 (which corresponds to log2 fold-change -0.585 and 0.585, respectively. A standard threshold for the adjusted p-value is less than 0.05.

Let's set these as variables to reuse. This is generally good practice because if these thresholds change upon later consideration, you only have to change them in one location of your script, which will reduce errors from missing some instances in your code.

```{r PlotSetup1}
fc = 1.5
fdr = 0.05
```

> # Note: Choosing thresholds {.unlisted .unnumbered}
>
> Thresholding on adjusted p-values < 0.05 is a standard threshold, but depending on the research question and/or how the results will be used, other thresholds might be reasonable.
>
> There is a nice [Biostar post that discusses choosing adjusted p-value thresholds](https://www.biostars.org/p/209118/), including cases where a more relaxed threshold might be appropriate and (some heated) discussion of the dangers of adjusting the choosen threshold after running an analysis. Additionally, there is a [Dalmon et al 2012](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-13-S2-S11) paper about p-value and fold-change thresholds for microarray data that may help provide some context.

To generate a general summary of the DE results, we can use the `summary` function to generate a basic summary by DESeq2.

```{r DESeq2Summary}
summary(results_minus_vs_plus)
```

However, this summary is simply a text output that we are unable to manipulate. Moreover, notice that the thresholds are not quite as we would like them.

We can use conditional statements to determine the number of genes that pass our thresholds for each comparison, which we can then use to add information to the results table and plots.

> # Exercise {.unlisted .unnumbered}
> How would we identify the number of genes with adjusted p-values < 0.05 and a fold-change above 1.5 (or below -1.5)?

<details>
<summary>Solution</summary>

Here is one possible answer:

```{r StatSigGenes2}
sum(results_minus_vs_plus$padj < fdr & abs(results_minus_vs_plus$log2FoldChange) >= log2(fc), na.rm = TRUE)
```
</details>
<br>

Let's now create a new column in `results_minus_vs_plus` to record the significance "call" based on these thresholds. And let's separate the call by "Up" or "Down", noting that these are relative to our "Case" condition of "plus", or iron replete mice. There are many ways to accomplish this, but the following will work:

First define all values as "NS" or "not significant":

```{r NSColumn}
results_minus_vs_plus$call = 'NS'
head(results_minus_vs_plus)
```

Next, determine the "Up" and "Down" indices:

```{r UpDownIndices}
up_idx = results_minus_vs_plus$padj < fdr & results_minus_vs_plus$log2FoldChange > log2(fc)
down_idx = results_minus_vs_plus$padj < fdr & results_minus_vs_plus$log2FoldChange < -log2(fc)
```

Last, use those indices to assign the correct "Up" or "Down" values to the correct indices, and look at the head of the result:

```{r CallColumn}
results_minus_vs_plus$call[up_idx] = 'Up'
results_minus_vs_plus$call[down_idx] = 'Down'
head(results_minus_vs_plus)
```

Finally, looking ahead to when we plot these values as colors in a volcano plot, let's make this `call` column a factor and specify the level ordering:

```{r FactorCall}
results_minus_vs_plus$call = factor(results_minus_vs_plus$call, levels = c('Up', 'Down', 'NS'))
head(results_minus_vs_plus)
```

> # Tip {.unlisted .unnumbered}
> It is often helpful to include code like this in differential expression analyses so there is a clearly labelled column that makes subsetting and summarizing the results easier.

Now we are in a position to quickly summarize our differential expression results:

```{r TableCall}
table(results_minus_vs_plus$call)
```

We see quickly how many genes were "Up" in iron replete, how many were "Down" in iron replete, and how many were not significant.

**Checkpoint**: *If you successfully added the `call` column and got the same table result as above, please indicate with a green check. Otherwise use a red x.*

## Visual DE summary

As described by this [Galaxy project tutorial](https://galaxyproject.github.io/training-material/topics/transcriptomics/tutorials/rna-seq-viz-with-volcanoplot/tutorial.html), a volcano plot is a type of scatterplot that shows statistical significance (adjusted p-value) versus magnitude of change (fold change). In a volcano plot, the most upregulated genes are towards the right, the most downregulated genes are towards the left, and the most statistically significant genes are towards the top.

Let's coerce the `DataFrame` which was returned by `DESeq2::results()` into a `tibble` in anticipation of using the `ggplot2` library to plot. We're also going to modify our results table so that the row names become a separate column, and so that it's ordered by adjusted p-value.

```{r PlotSetup2}
# Use the rownames argument to create a new column of gene IDs
# Also arrange by adjusted p-value
results_minus_vs_plus = as_tibble(results_minus_vs_plus, rownames = 'id') %>% arrange(padj)
```

Let's start with a very simple volcano plot that plots the `log2FoldChange` on the x-axis, and `-log10(padj)` on the y-axis.

```{r VolcanoPlot}
# Initialize the plot, saving as object 'p' and specifying the plot type as 'geom_point'
p = ggplot(results_minus_vs_plus, aes(x = log2FoldChange, y = -log10(padj))) +
    geom_point()
p
```

This is a good start, but, as usual it's nice to add better labels to the plot with the `labs()` function:

```{r VolcanoPlot2}
# Add plot labels and change the theme
p = ggplot(results_minus_vs_plus, aes(x = log2FoldChange, y = -log10(padj))) +
    geom_point() +
    theme_bw() +
    labs(
        title = 'Volcano Plot',
        subtitle = 'Plus vs Minus',
        x = 'log2 fold-change',
        y = '-log10 FDR'
    )
p
```

This is a nice improvement. What if we now added some visual guides to indicate where the significant genes are? We can uset he `geom_vline()` and `geom_hline()` functions to accomplish this:

```{r VolcanoPlot3}
# Add threshold lines
p = p +
    geom_vline(
        xintercept = c(0, -log2(fc), log2(fc)),
        linetype = c(1, 2, 2)) +
    geom_hline(
        yintercept = -log10(fdr),
        linetype = 2)
p
```

Finally, why not color the points by their significance status? We already created the `call` column that has the correct values. In this case we can get away with adding a last `geom_point()` and specifying the correct aesthetic:

```{r VolcanoPlot4}
p = p + geom_point(aes(color = call))
p
```

For additional visualizations for our DE results, we included some example code in the Bonus Content module and this [HBC tutorial](https://hbctraining.github.io/DGE_workshop/lessons/06_DGE_visualizing_results.html) includes some nice examples.

### Subsetting significant genes

You may be interested in identifying only the genes that pass your significance thresholds. A useful way to do this is to conditionally subset your results. Again, we already created the `call` column, which makes this relatively simple to do.

*Note: The tidyverse functions you learned in Software Carpentry could also be alternatively used here.*

```{r ConditionalSubset}
res_sig = results_minus_vs_plus[results_minus_vs_plus$call != 'NS', ]
head(res_sig)
dim(res_sig)
```

## Adding genome annotations

Since, gene symbols can change over time or be ambiguous we use, and recommend, using the EMSEMBL reference genome and ENSEMBL IDs for alignments and we've been working with tables and data where all genes are labeled only by their long ENSEMBL ID (you will notice this in the GTF we used in the reference). However, this can make it difficult to quickly look at results for genes of interest.

Luckily, Bioconductor provides many tools and resources to facilitate access to [genomic annotation resources](http://bioconductor.org/packages/devel/workflows/vignettes/annotation/inst/doc/Annotation_Resources.html).

To start, we will first load the [biomaRt library](https://bioconductor.org/packages/3.14/bioc/html/biomaRt.html) and choose what reference we want to access. For a more detailed walk through of using biomaRt, [this training module](https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2019/RNAseq/html/05_Annotation_and_Visualisation.html) might be useful, including what to do when annotations are not 1:1 mappings.

We'll start by loading the `biomaRt` library and calling the `useEnsembl()` function to select the database we'll use to extract the information we need. This will download the mapping of ENSEMBL IDs to gene symbols, enabling us to eventually add the gene symbol column we want.

```{r Pullmart, warning=FALSE}
library('biomaRt')
ensembl = useEnsembl(dataset = 'mmusculus_gene_ensembl', biomart='ensembl')
```

*Note that this process takes some time and will take up a larger amount of working memory so proceed with caution if you try to run these commands on a laptop with less than 4G of memory*

To identify possible **filters** to restrict our data, we can use the `listFilters` function. To identify the **attributes** we want to retrive, we can use the `listAttributes` function. The best approach is to use [list or search functions](https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html#how-to-build-a-biomart-query) to help narrow down the available options.

```{r AddAnnotations2, warning=FALSE, eval=FALSE}
head(listFilters(mart = ensembl), n = 20)
head(listAttributes(ensembl), n = 30)
```

We can access additional genomic annotations using the [`bioMart` package](https://bioconductor.org/packages/release/bioc/html/biomaRt.html). To identify  we'll structure our 'query' or search of the bioMart resources to use the [ENSEMBL id](https://m.ensembl.org/info/genome/genebuild/gene_names.html) from our alignment to add the gene symbols and gene description for each gene.

```{r AddAnotation3, warning=FALSE}
id_mapping = getBM(attributes=c('ensembl_gene_id', 'external_gene_name'),
      filters = 'ensembl_gene_id',
      values = row.names(assay(dds)),
      mart = ensembl)
# will take some time to run

# Preview the result
head(id_mapping)
```

Now that we have the ENSEMBL information and a gene symbol to match to our results, we can proceed in the smaller groups. As with the previous exercise, we have broken it into small steps with hints as needed.

**Note**: For additional information regarding bioMart, please consult the [ENSEMBL bioMart vignette](https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html) or the broader [Bioconductor Annotation Resources vignette ](http://bioconductor.org/packages/devel/workflows/vignettes/annotation/inst/doc/Annotation_Resources.html).

Look at the two data frames that are going to be needed: `id_mapping` and `results_minus_vs_plus`.

```{r preview_tables}
head(id_mapping)
head(results_minus_vs_plus)
```

We want to match the `id` column of `results_minus_vs_plus` to the `ensembl_gene_id` column of `id_mapping`, and once that match is found, we want to extract the `external_gene_name` column of `id_mapping` to get the gene symbol. Next, look at the documentation for `dplyr::left_join()` and merge the `id_mapping` table into the `results_minus_vs_plus` table on the columns `ensembl_gene_id` and `external_gene_name`.

```{r left_join}
results_minus_vs_plus_annotated = results_minus_vs_plus %>%
    left_join(id_mapping, by = c('id' = 'ensembl_gene_id'))
head(results_minus_vs_plus_annotated)
```

We can use some of the `tidyverse` functions we've encountered previously to rename the `external_gene_name` column to `symbol` and to move it into the second column position? Hint: Because of the order of the packages we may have loaded, we'll use `dplyr::rename()` and `dplyr::select()` instead of just the `select()` function. We can discuss this in a moment.

```{r rename_rearrange}
results_minus_vs_plus_annotated = results_minus_vs_plus_annotated %>%
    dplyr::rename('symbol' = 'external_gene_name') %>%
    dplyr::select(id, symbol, everything())
results_minus_vs_plus_annotated
```

And now we have our differential expression results annotated with gene symbols, which can help in the interpretation of the results, and can be used in downstream analysis such as functional analysis.

# Outputting results to file

A key aspect of our analysis is preserving the relevant datasets for both our records and for downstream applications, such as functional enrichments.


## DE results table

Next we'll write out our DE results, now that we've added information to the table to help us interpret the results, and share with collaborators.

```{r DEResultsOuput, eval = FALSE}
write.csv(results_minus_vs_plus,
          row.names = FALSE,
          na = ".",
          file="outputs/tables/DE_results_minus_vs_plus.csv")
```

## Count tables

The most relevant count tables are the raw, filtered count table that we used as the input for our analysis and the rlog normalized count table that we used for our quality control visualizations.

To output the raw counts, we will need to use the `counts` function to access the count table from within its larger `DESeqDataSet` object.

```{r OutputCountsRaw, eval = FALSE}
write.csv(counts(dds, normalized = FALSE), file="outputs/tables/DESeq2_raw_counts.csv")
```

Then we'll output the rlog count table, using the `assay` function to access the normalized count table from within its larger `DESeqDataSet` object.

```{r OutputCountsRlog, eval = FALSE}
write.csv(assay(rld), file="outputs/tables/DESeq2_rlog_normalized_counts.csv")
```

# Summary

In this section, we:

* Generated a volcano plot for our differential expression results
* Summarized our differential expression results
* Discussed choosing thresholds
* Annotated our tables of results to map gene IDs to gene symbols
* Saved our results to file


# Key takeaways

Overall, we've run through most of the building blocks needed to run a differential expression analysis and hopefully built up a better understanding of how differential expression comparisons work, particularly how experimental design can impact our results.

What to consider moving forward:

* How can I control for technical variation in my experimental design?
* How much variation is expected with a treatment group?
* What is my RNA quality, and how can that be optimized?
* Are there quality concerns for my sequencing data?
* What comparisons are relevant to my biological question?
* Are there covariates that should be considered?
* What will a differential expression analysis tell me?


**Let's pause here for general questions**

---

# What can we do with our DE results?

A way to determine possible [broader biological interpretations](https://www.ebi.ac.uk/training-beta/online/courses/functional-genomics-ii-common-technologies-and-data-analysis-methods/biological-interpretation-of-gene-expression-data-2/) from the observed DE results, is functional enrichments. There are many options, such as some included in this [discussion thread](https://www.researchgate.net/post/How_can_I_analyze_a_set_of_DEGs_differentially_expressed_genes_to_obtain_information_from_them). Other common functional enrichments approaches are gene set enrichment analysis, aka [GSEA](http://software.broadinstitute.org/gsea/index.jsp), Database for Annotation, Visualization and Integrated Discovery, aka [DAVID](https://david.ncifcrf.gov/), [Ingenity](https://digitalinsights.qiagen.com/), and [iPathway Guide]

The University of Michigan has license and support for additional tools, such as Cytoscape, so we recommend reaching out to staff with [Taubman Library](https://www.lib.umich.edu/locations-and-hours/taubman-health-sciences-library/research-and-clinical-support) to learn more about resources that might be application toyour research.

---

# Sources

* HBC DGE training module, part 1: https://hbctraining.github.io/DGE_workshop/lessons/04_DGE_DESeq2_analysis.html
* HBC DGE training module, part 2: https://hbctraining.github.io/DGE_workshop/lessons/05_DGE_DESeq2_analysis2.html
* DESeq2 vignette: http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#differential-expression-analysis
* Bioconductor Genomic Annotation resources: http://bioconductor.org/packages/devel/workflows/vignettes/annotation/inst/doc/Annotation_Resources.html
* BioMart vignette: https://bioconductor.org/packages/release/bioc/vignettes/biomaRt/inst/doc/accessing_ensembl.html

# Additional Resources
* MIDAS Reproduciblity Hub: https://midas.umich.edu/reproducibility-overview/
* ARC resources: https://arc-ts.umich.edu/
* Gene Set Enrichment Resources from Bioconductor: https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2018/RNASeq2018/html/06_Gene_set_testing.nb.html
* Using HTSeq data with DESeq2: https://angus.readthedocs.io/en/2019/diff-ex-and-viz.html
* Detailed RNA-seq analysis paper: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6096346/
* Overview of RNA-seq analysis considerations: https://academic-oup-com.proxy.lib.umich.edu/bfg/article/14/2/130/257370
* Alternative overview of DESeq2, including visualizations and functional enrichments: http://dputhier.github.io/jgb71e-polytech-bioinfo-app/practical/rna-seq_R/rnaseq_diff_Snf2.html

---

```{r WriteOut.RData, eval=TRUE, echo=FALSE, message=FALSE, warning=FALSE}
# Hidden code block to write out data for knitting
# save.image(file = "rdata/RunningData_Full.RData")
```

# Session Info
```{r SessionInfo}
sessionInfo()
```

---

These materials have been adapted and extended from materials listed above. These are open access materials distributed under the terms of the [Creative Commons Attribution license (CC BY 4.0)](http://creativecommons.org/licenses/by/4.0/), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeSourceEmbed("Module11_DEVisualizations.Rmd");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
